ipa-core/src/protocol/basics/mul/semi_honest.rs

```diff
@@ -182,4 +201,145 @@ mod test {
+                multiply(
+                    ctx.set_total_records(1),
+                    RecordId::from(0),
+                    &a_shares,
+                    &b_shares,
+                    ZeroPositions::NONE,
+                )
+                .await
+                .unwrap()
+            })
+            .await;
+        assert_eq!(expected, results.reconstruct_arr());
+    }
+
+    // The manymult test is a microbenchmark. The test generates a DxW matrix of field elements. The
+    // matrix is reduced to a single W-element row vector by taking the element-wise product of the
+    // D values in each column. The non-vectorized implementation (manymult_novec) simply does a
+    // parallel_join of W semi-honest multiplies. The vectorized implementation (manymult_vec)
+    // processes a row at a time. For manymult_vec, MANYMULT_WIDTH must match a supported
+    // vectorization width.
+    const MANYMULT_ITERS: usize = 512;
+    const MANYMULT_WIDTH: usize = 32;
+
+    #[tokio::test]
+    pub async fn manymult_novec() {
+        let world = TestWorld::default();
+        let mut rng = thread_rng();
+        let mut inputs = Vec::<Vec<Fp32BitPrime>>::new();
+        for _ in 0..MANYMULT_ITERS {
+            inputs.push(
+                (0..MANYMULT_WIDTH)
+                    .map(|_| Fp32BitPrime::try_from(u128::from(rng.gen_range(0u32..100))).unwrap())
+                    .collect::<Vec<_>>(),
+            );
+        }
+        let expected = inputs
+            .iter()
+            .fold(None, |acc: Option<Vec<Fp32BitPrime>>, b| match acc {
+                Some(a) => Some(a.iter().zip(b.iter()).map(|(&a, &b)| a * b).collect()),
+                None => Some(b.clone()),
+            })
+            .unwrap();
+
+        let begin = Instant::now();
+        let result = world
+            .semi_honest(
+                inputs.into_iter().map(IntoIterator::into_iter),
+                |ctx, share: Vec<Vec<AdditiveShare<Fp32BitPrime>>>| async move {
+                    let ctx = ctx.set_total_records(MANYMULT_ITERS * MANYMULT_WIDTH);
+                    let mut iter = share.iter();
+                    let mut val = iter.next().unwrap().clone();
+                    for i in 1..MANYMULT_ITERS {
+                        let cur = iter.next().unwrap();
+                        let mut res = Vec::with_capacity(MANYMULT_WIDTH);
+                        for j in 0..MANYMULT_WIDTH {
+                            res.push(val[j].multiply(
+                                &cur[j],
+                                ctx.clone(),
+                                RecordId::from(MANYMULT_WIDTH * (i - 1) + j),
+                            ));
+                        }
+                        val = ctx.parallel_join(res).await.unwrap();
+                    }
+                    val
+                },
+            )
+            .await;
+        tracing::debug!("Protocol execution time: {:?}", begin.elapsed());
```

Comment:
we should benchmark it again after #873 is merged

Reply:
I modified the arithmetic circuit benchmark to use seq_join, and ran a bunch configurations of the modified version. It seems like parallelization does not help this benchmark much (either with or without vectorization), and in some cases even makes it slightly slower.

[vectorization-perf-2.txt](https://github.com/private-attribution/ipa/files/14103387/vectorization-perf-2.txt)

(In all configurations, these results have PRSS used index tracking disabled, which gives a ~2x speed up compared to the results in the description. Increasing the vectorization width from 32 to 256 gives another ~50% on top of that.)

Reply:
It looks like this test case isn't using metrics because it is not using `TestWorld::semi_honest` or `TestWorld::malicious` (it is using `TestWorld`, but not the convenience wrapper for sharing/reconstructing I/O, which is what registers a tracing span). I wasn't running with `disable-metrics`, but adding it doesn't change the results. Turning off stall detection improves performance 5-10% in the novec-par case, doesn't make much difference in the vec-par case.

Reply:
If turning off stall detection helps, it could be explained by contention on accessing send and receive channels.